In [3]:
import sqlite3
import pandas as pd
conn = sqlite3.connect(f"{base_folder}/data/housing.db")
housing = pd.read_sql_query(
    """
    SELECT
        b.block_id,
        b.longitude,
        b.latitude,
        s.housing_median_age,
        s.total_rooms,
        s.total_bedrooms,
        s.population,
        s.households,
        s.median_income,
        s.median_house_value,
        op.name AS ocean_proximity
    FROM block AS b
    JOIN block_housing_stats AS s
        ON s.block_id = b.block_id
    JOIN ocean_proximity AS op
        ON op.ocean_proximity_id = b.ocean_proximity_id
    ORDER BY b.block_id
    """,
    conn,
)
conn.close()

housing.head()

block_id  longitude  latitude  housing_median_age  total_rooms  \
0         0    -122.23     37.88                41.0          880   
1         1    -122.22     37.86                21.0         7099   
2         2    -122.24     37.85                52.0         1467   
3         3    -122.25     37.85                52.0         1274   
4         4    -122.25     37.85                52.0         1627   

   total_bedrooms  population  households  median_income  median_house_value  \
0           129.0         322         126         8.3252            452600.0   
1          1106.0        2401        1138         8.3014            358500.0   
2           190.0         496         177         7.2574            352100.0   
3           235.0         558         219         5.6431            341300.0   
4           280.0         565         259         3.8462            342200.0   

  ocean_proximity  
0        NEAR BAY  
1        NEAR BAY  
2        NEAR BAY  
3        NEAR BAY  
4        NEAR BAY

In [4]:
# =============================================================================
# FULL PIPELINE with OPTUNA
# - Build preprocessing
# - Stratified train/test split
# - Train & log 4 models WITHOUT PCA (Ridge, HGB, XGBoost, LightGBM)
# - Train & log 4 models WITH PCA (preprocessing + PCA(0.95) + model)
# - Pick GLOBAL best among 8 models by Test MAE
# - Save, load, and compare the global best model
# =============================================================================

import time
import os
import numpy as np
import pandas as pd

from dotenv import load_dotenv

from sklearn.compose import ColumnTransformer
from sklearn.decomposition import PCA
from sklearn.impute import SimpleImputer
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.pipeline import Pipeline, make_pipeline
from sklearn.preprocessing import FunctionTransformer, OneHotEncoder, StandardScaler
from sklearn.linear_model import Ridge
from sklearn.ensemble import HistGradientBoostingRegressor

import mlflow
from mlflow.models import infer_signature
import joblib

from xgboost import XGBRegressor
from lightgbm import LGBMRegressor

import optuna
from optuna.samplers import TPESampler

from sklearn.base import clone

# Shared components
from housing_pipeline import (
    build_preprocessing,
    make_estimator_for_name,
)

start_time = time.monotonic()
optuna.logging.set_verbosity(optuna.logging.WARNING)


# =============================================================================
# STEP 1: Build Full ML Preprocessing Pipeline
# =============================================================================

preprocessing = build_preprocessing()
print("✓ STEP 1: Preprocessing pipeline created.")


# =============================================================================
# STEP 2: Split Data into Stratified Train and Test Sets
# =============================================================================

housing["income_cat"] = pd.cut(
    housing["median_income"],
    bins=[0, 1.5, 3.0, 4.5, 6, np.inf],
    labels=[1, 2, 3, 4, 5],
)

train_set, test_set = train_test_split(
    housing,
    test_size=0.20,
    stratify=housing["income_cat"],
    random_state=42,
)

for df in (train_set, test_set):
    df.drop("income_cat", axis=1, inplace=True)

X_train = train_set.drop(["block_id", "median_house_value"], axis=1).copy()
y_train = train_set["median_house_value"].copy()

X_test = test_set.drop(["block_id", "median_house_value"], axis=1).copy()
y_test = test_set["median_house_value"].copy()

print(f"✓ STEP 2: Stratified split done. Train size: {len(X_train)}, Test size: {len(X_test)}")


# =============================================================================
# STEP 3: Configure MLflow
# =============================================================================

load_dotenv(
    dotenv_path="/content/drive/MyDrive/Colab Notebooks/housing_fall2025/notebooks/.env",
    override=True
)

MLFLOW_TRACKING_URI = os.getenv("MLFLOW_TRACKING_URI")
MLFLOW_TRACKING_USERNAME = os.getenv("MLFLOW_TRACKING_USERNAME")
MLFLOW_TRACKING_PASSWORD = os.getenv("MLFLOW_TRACKING_PASSWORD")

if MLFLOW_TRACKING_USERNAME:
    os.environ["MLFLOW_TRACKING_USERNAME"] = MLFLOW_TRACKING_USERNAME
if MLFLOW_TRACKING_PASSWORD:
    os.environ["MLFLOW_TRACKING_PASSWORD"] = MLFLOW_TRACKING_PASSWORD

mlflow.set_tracking_uri(MLFLOW_TRACKING_URI)
mlflow.set_experiment("median_house_pricing_multi_model_optuna")

print("✓ STEP 3: MLflow configured.")


# =============================================================================
# STEP 4: Define Optuna Objective Functions (NO PCA)
# =============================================================================

def objective_ridge(trial, preprocessing, X_train, y_train):
    alpha = trial.suggest_float("ridge__alpha", 0.1, 100.0, log=True)
    preprocessing_clone = clone(preprocessing)
    pipeline = make_pipeline(preprocessing_clone, Ridge(alpha=alpha))
    scores = cross_val_score(
        pipeline, X_train, y_train,
        cv=3, scoring="neg_mean_absolute_error", n_jobs=-1
    )
    return -scores.mean()


def objective_hgb(trial, preprocessing, X_train, y_train):
    learning_rate = trial.suggest_float("hgb__learning_rate", 0.05, 0.2)
    max_depth = trial.suggest_int("hgb__max_depth", 3, 8)
    preprocessing_clone = clone(preprocessing)
    pipeline = make_pipeline(
        preprocessing_clone,
        HistGradientBoostingRegressor(
            learning_rate=learning_rate,
            max_depth=max_depth,
            random_state=42
        )
    )
    scores = cross_val_score(
        pipeline, X_train, y_train,
        cv=3, scoring="neg_mean_absolute_error", n_jobs=-1
    )
    return -scores.mean()


def objective_xgb(trial, preprocessing, X_train, y_train):
    learning_rate = trial.suggest_float("xgb__learning_rate", 0.05, 0.2)
    max_depth = trial.suggest_int("xgb__max_depth", 3, 8)
    n_estimators = trial.suggest_int("xgb__n_estimators", 100, 300, step=50)
    preprocessing_clone = clone(preprocessing)
    pipeline = make_pipeline(
        preprocessing_clone,
        XGBRegressor(
            objective="reg:squarederror",
            random_state=42,
            n_estimators=n_estimators,
            learning_rate=learning_rate,
            max_depth=max_depth,
            tree_method="hist",
            n_jobs=-1,
        )
    )
    scores = cross_val_score(
        pipeline, X_train, y_train,
        cv=3, scoring="neg_mean_absolute_error", n_jobs=-1
    )
    return -scores.mean()


def objective_lgbm(trial, preprocessing, X_train, y_train):
    learning_rate = trial.suggest_float("lgbm__learning_rate", 0.05, 0.2)
    num_leaves = trial.suggest_int("lgbm__num_leaves", 20, 80)
    n_estimators = trial.suggest_int("lgbm__n_estimators", 100, 300, step=50)
    preprocessing_clone = clone(preprocessing)
    pipeline = make_pipeline(
        preprocessing_clone,
        LGBMRegressor(
            random_state=42,
            n_estimators=n_estimators,
            learning_rate=learning_rate,
            num_leaves=num_leaves,
            n_jobs=-1,
            verbose=-1,
        )
    )
    scores = cross_val_score(
        pipeline, X_train, y_train,
        cv=3, scoring="neg_mean_absolute_error", n_jobs=-1
    )
    return -scores.mean()


# =============================================================================
# STEP 5: Run Optuna Studies for Each Model (NO PCA)
# =============================================================================

model_names = ["ridge", "histgradientboosting", "xgboost", "lightgbm"]
objective_functions = {
    "ridge": objective_ridge,
    "histgradientboosting": objective_hgb,
    "xgboost": objective_xgb,
    "lightgbm": objective_lgbm,
}

results = {}

for name in model_names:
    print(f"\n{'='*80}")
    print(f"Optimizing {name.upper()} (NO PCA) - 10 trials")
    print(f"{'='*80}")

    study = optuna.create_study(
        direction="minimize",
        sampler=TPESampler(seed=42),
        study_name=f"{name}_study"
    )

    study.optimize(
        lambda trial: objective_functions[name](trial, preprocessing, X_train, y_train),
        n_trials=10,
        show_progress_bar=True
    )

    cv_mae = study.best_value
    print(f"\nBest {name.upper()} CV MAE: ${cv_mae:,.2f}")
    print(f"Best params: {study.best_params}")

    best_params = study.best_params
    preprocessing_clone = clone(preprocessing)

    if name == "ridge":
        final_model = make_pipeline(
            preprocessing_clone,
            Ridge(alpha=best_params["ridge__alpha"])
        )
    elif name == "histgradientboosting":
        final_model = make_pipeline(
            preprocessing_clone,
            HistGradientBoostingRegressor(
                learning_rate=best_params["hgb__learning_rate"],
                max_depth=best_params["hgb__max_depth"],
                random_state=42
            )
        )
    elif name == "xgboost":
        final_model = make_pipeline(
            preprocessing_clone,
            XGBRegressor(
                objective="reg:squarederror",
                random_state=42,
                n_estimators=best_params["xgb__n_estimators"],
                learning_rate=best_params["xgb__learning_rate"],
                max_depth=best_params["xgb__max_depth"],
                tree_method="hist",
                n_jobs=-1,
            )
        )
    elif name == "lightgbm":
        final_model = make_pipeline(
            preprocessing_clone,
            LGBMRegressor(
                random_state=42,
                n_estimators=best_params["lgbm__n_estimators"],
                learning_rate=best_params["lgbm__learning_rate"],
                num_leaves=best_params["lgbm__num_leaves"],
                n_jobs=-1,
                verbose=-1,
            )
        )

    final_model.fit(X_train, y_train)

    y_pred = final_model.predict(X_test)
    test_mae = mean_absolute_error(y_test, y_pred)
    print(f"{name} (no PCA) Test MAE: ${test_mae:,.2f}")

    results[name] = {"pipeline": final_model, "test_mae": test_mae, "cv_mae": cv_mae}

    with mlflow.start_run(run_name=f"{name}_baseline_optuna"):
        mlflow.log_param("model_family", name)
        mlflow.log_param("uses_pca", False)
        mlflow.log_params(best_params)
        mlflow.log_metric("cv_MAE", cv_mae)
        mlflow.log_metric("test_MAE", test_mae)

        signature = infer_signature(X_train, final_model.predict(X_train))
        mlflow.sklearn.log_model(
            sk_model=final_model,
            artifact_path="housing_model",
            signature=signature,
            input_example=X_train,
            registered_model_name=f"{name}_pipeline_optuna",
        )

print("\n✓ STEP 5: All 4 baseline models optimized and logged.")


# =============================================================================
# STEP 6: PCA Optuna Objectives
# =============================================================================

def objective_ridge_pca(trial, preprocessing, X_train, y_train):
    alpha = trial.suggest_float("ridge__alpha", 0.1, 100.0, log=True)
    pca_components = trial.suggest_float("pca__n_components", 0.90, 0.99)
    preprocessing_clone = clone(preprocessing)
    pipeline = make_pipeline(preprocessing_clone, PCA(n_components=pca_components), Ridge(alpha=alpha))
    scores = cross_val_score(
        pipeline, X_train, y_train,
        cv=3, scoring="neg_mean_absolute_error", n_jobs=-1
    )
    return -scores.mean()


def objective_hgb_pca(trial, preprocessing, X_train, y_train):
    learning_rate = trial.suggest_float("hgb__learning_rate", 0.05, 0.2)
    max_depth = trial.suggest_int("hgb__max_depth", 3, 8)
    pca_components = trial.suggest_float("pca__n_components", 0.90, 0.99)
    preprocessing_clone = clone(preprocessing)
    pipeline = make_pipeline(
        preprocessing_clone,
        PCA(n_components=pca_components),
        HistGradientBoostingRegressor(
            learning_rate=learning_rate,
            max_depth=max_depth,
            random_state=42
        )
    )
    scores = cross_val_score(
        pipeline, X_train, y_train,
        cv=3, scoring="neg_mean_absolute_error", n_jobs=-1
    )
    return -scores.mean()


def objective_xgb_pca(trial, preprocessing, X_train, y_train):
    learning_rate = trial.suggest_float("xgb__learning_rate", 0.05, 0.2)
    max_depth = trial.suggest_int("xgb__max_depth", 3, 8)
    n_estimators = trial.suggest_int("xgb__n_estimators", 100, 300, step=50)
    pca_components = trial.suggest_float("pca__n_components", 0.90, 0.99)
    preprocessing_clone = clone(preprocessing)
    pipeline = make_pipeline(
        preprocessing_clone,
        PCA(n_components=pca_components),
        XGBRegressor(
            objective="reg:squarederror",
            random_state=42,
            n_estimators=n_estimators,
            learning_rate=learning_rate,
            max_depth=max_depth,
            tree_method="hist",
            n_jobs=-1,
        )
    )
    scores = cross_val_score(
        pipeline, X_train, y_train,
        cv=3, scoring="neg_mean_absolute_error", n_jobs=-1
    )
    return -scores.mean()


def objective_lgbm_pca(trial, preprocessing, X_train, y_train):
    learning_rate = trial.suggest_float("lgbm__learning_rate", 0.05, 0.2)
    num_leaves = trial.suggest_int("lgbm__num_leaves", 20, 80)
    n_estimators = trial.suggest_int("lgbm__n_estimators", 100, 300, step=50)
    pca_components = trial.suggest_float("pca__n_components", 0.90, 0.99)
    preprocessing_clone = clone(preprocessing)
    pipeline = make_pipeline(
        preprocessing_clone,
        PCA(n_components=pca_components),
        LGBMRegressor(
            random_state=42,
            n_estimators=n_estimators,
            learning_rate=learning_rate,
            num_leaves=num_leaves,
            n_jobs=-1,
            verbose=-1,
        )
    )
    scores = cross_val_score(
        pipeline, X_train, y_train,
        cv=3, scoring="neg_mean_absolute_error", n_jobs=-1
    )
    return -scores.mean()


# =============================================================================
# STEP 7: Run Optuna Studies for PCA Models
# =============================================================================

pca_model_names = ["ridge_with_pca", "histgradientboosting_with_pca", "xgboost_with_pca", "lightgbm_with_pca"]
pca_objective_functions = {
    "ridge_with_pca": objective_ridge_pca,
    "histgradientboosting_with_pca": objective_hgb_pca,
    "xgboost_with_pca": objective_xgb_pca,
    "lightgbm_with_pca": objective_lgbm_pca,
}

pca_results = {}

for name in pca_model_names:
    base_name = name.replace("_with_pca", "")
    print(f"\n{'='*80}")
    print(f"Optimizing {name.upper()} - 10 trials")
    print(f"{'='*80}")

    study = optuna.create_study(
        direction="minimize",
        sampler=TPESampler(seed=42),
        study_name=f"{name}_study"
    )

    study.optimize(
        lambda trial: pca_objective_functions[name](trial, preprocessing, X_train, y_train),
        n_trials=10,
        show_progress_bar=True
    )

    cv_mae_pca = study.best_value
    print(f"\nBest {name.upper()} CV MAE: ${cv_mae_pca:,.2f}")
    print(f"Best params: {study.best_params}")

    best_params = study.best_params
    preprocessing_clone = clone(preprocessing)

    if base_name == "ridge":
        final_model = make_pipeline(
            preprocessing_clone,
            PCA(n_components=best_params["pca__n_components"]),
            Ridge(alpha=best_params["ridge__alpha"])
        )
    elif base_name == "histgradientboosting":
        final_model = make_pipeline(
            preprocessing_clone,
            PCA(n_components=best_params["pca__n_components"]),
            HistGradientBoostingRegressor(
                learning_rate=best_params["hgb__learning_rate"],
                max_depth=best_params["hgb__max_depth"],
                random_state=42
            )
        )
    elif base_name == "xgboost":
        final_model = make_pipeline(
            preprocessing_clone,
            PCA(n_components=best_params["pca__n_components"]),
            XGBRegressor(
                objective="reg:squarederror",
                random_state=42,
                n_estimators=best_params["xgb__n_estimators"],
                learning_rate=best_params["xgb__learning_rate"],
                max_depth=best_params["xgb__max_depth"],
                tree_method="hist",
                n_jobs=-1,
            )
        )
    elif base_name == "lightgbm":
        final_model = make_pipeline(
            preprocessing_clone,
            PCA(n_components=best_params["pca__n_components"]),
            LGBMRegressor(
                random_state=42,
                n_estimators=best_params["lgbm__n_estimators"],
                learning_rate=best_params["lgbm__learning_rate"],
                num_leaves=best_params["lgbm__num_leaves"],
                n_jobs=-1,
                verbose=-1,
            )
        )

    final_model.fit(X_train, y_train)

    y_pred = final_model.predict(X_test)
    test_mae = mean_absolute_error(y_test, y_pred)
    print(f"{name} Test MAE: ${test_mae:,.2f}")

    pca_results[name] = {"pipeline": final_model, "test_mae": test_mae, "cv_mae": cv_mae_pca}

    with mlflow.start_run(run_name=f"{name}_optuna"):
        mlflow.log_param("model_family", base_name)
        mlflow.log_param("uses_pca", True)
        mlflow.log_params(best_params)
        mlflow.log_metric("cv_MAE", cv_mae_pca)
        mlflow.log_metric("test_MAE", test_mae)

        signature = infer_signature(X_train, final_model.predict(X_train))
        mlflow.sklearn.log_model(
            sk_model=final_model,
            artifact_path="housing_model_with_pca",
            signature=signature,
            input_example=X_train,
            registered_model_name=f"{base_name}_pipeline_with_pca_optuna",
        )

print("\n✓ STEP 7: All 4 PCA models optimized and logged.")


# =============================================================================
# STEP 8: Choose GLOBAL Best Model
# =============================================================================

all_results = {}
all_results.update(results)
all_results.update(pca_results)

global_best_name = min(all_results, key=lambda k: all_results[k]["test_mae"])
global_best_mae = all_results[global_best_name]["test_mae"]
global_best_cv_mae = all_results[global_best_name]["cv_mae"]
global_best_pipeline = all_results[global_best_name]["pipeline"]

uses_pca = "with_pca" in global_best_name

print("\n" + "=" * 80)
print("GLOBAL BEST MODEL (ACROSS 8 CANDIDATES)")
print("=" * 80)
print(f"Global best model key: {global_best_name}")
print(f"Global best CV MAE:    ${global_best_cv_mae:,.2f}")
print(f"Global best Test MAE:  ${global_best_mae:,.2f}")
print(f"Uses PCA:               {uses_pca}")


# =============================================================================
# STEP 9: Save, Load, and Compare the GLOBAL Best Model
# =============================================================================

def save_model(model, filename="global_best_model_optuna.pkl"):
    joblib.dump(model, filename)
    print(f"✓ Model saved to {filename}")

print("\n" + "-" * 80)
print("Saving and reloading GLOBAL best model...")
print("-" * 80)

save_model(global_best_pipeline, filename=f"{base_folder}/models/global_best_model_optuna.pkl")

print("\nDone:")
print(f"- GLOBAL best model key: {global_best_name}")
print(f"- GLOBAL best CV MAE:    ${global_best_cv_mae:,.2f}")
print(f"- GLOBAL best Test MAE:  ${global_best_mae:,.2f}")

end_time = time.monotonic()
elapsed_time = end_time - start_time
minutes = int(elapsed_time // 60)
seconds = elapsed_time % 60
print(f"Elapsed time: {minutes} minutes and {seconds:.2f} seconds")

✓ STEP 1: Preprocessing pipeline created.
✓ STEP 2: Stratified split done. Train size: 16512, Test size: 4128
✓ STEP 3: MLflow configured.

Optimizing RIDGE (NO PCA) - 10 trials


  0%|          | 0/10 [00:00<?, ?it/s]


Best RIDGE CV MAE: $51,138.10
Best params: {'ridge__alpha': 0.14936568554617632}
ridge (no PCA) Test MAE: $52,372.31


/usr/local/lib/python3.12/dist-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
Registered model 'ridge_pipeline_optuna' already exists. Creating a new version of this model...
2025/12/11 11:50:26 INFO mlflow.store.model_registry.abstract_store: Waiting up to 300 seconds for model version to finish creation. Model name: ridge_

🏃 View run ridge_baseline_optuna at: https://dagshub.com/mkzia/housing_fall2025_1.mlflow/#/experiments/1/runs/7eee282079314d66b1271847962db2e8
🧪 View experiment at: https://dagshub.com/mkzia/housing_fall2025_1.mlflow/#/experiments/1

Optimizing HISTGRADIENTBOOSTING (NO PCA) - 10 trials


  0%|          | 0/10 [00:00<?, ?it/s]


Best HISTGRADIENTBOOSTING CV MAE: $30,313.48
Best params: {'hgb__learning_rate': 0.14016725176148134, 'hgb__max_depth': 7}
histgradientboosting (no PCA) Test MAE: $30,231.39


/usr/local/lib/python3.12/dist-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
Registered model 'histgradientboosting_pipeline_optuna' already exists. Creating a new version of this model...
2025/12/11 11:51:03 INFO mlflow.store.model_registry.abstract_store: Waiting up to 300 seconds for model version to finish creation. Mod

🏃 View run histgradientboosting_baseline_optuna at: https://dagshub.com/mkzia/housing_fall2025_1.mlflow/#/experiments/1/runs/65ab9012e39247adba5725f51110a507
🧪 View experiment at: https://dagshub.com/mkzia/housing_fall2025_1.mlflow/#/experiments/1

Optimizing XGBOOST (NO PCA) - 10 trials


  0%|          | 0/10 [00:00<?, ?it/s]


Best XGBOOST CV MAE: $28,986.41
Best params: {'xgb__learning_rate': 0.10618101782710439, 'xgb__max_depth': 8, 'xgb__n_estimators': 250}
xgboost (no PCA) Test MAE: $28,384.85


/usr/local/lib/python3.12/dist-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
Registered model 'xgboost_pipeline_optuna' already exists. Creating a new version of this model...
2025/12/11 11:52:12 INFO mlflow.store.model_registry.abstract_store: Waiting up to 300 seconds for model version to finish creation. Model name: xgbo

🏃 View run xgboost_baseline_optuna at: https://dagshub.com/mkzia/housing_fall2025_1.mlflow/#/experiments/1/runs/cfc252a5f5b346af88bff02cabff5e72
🧪 View experiment at: https://dagshub.com/mkzia/housing_fall2025_1.mlflow/#/experiments/1

Optimizing LIGHTGBM (NO PCA) - 10 trials


  0%|          | 0/10 [00:00<?, ?it/s]


Best LIGHTGBM CV MAE: $28,892.66
Best params: {'lgbm__learning_rate': 0.05871254182522992, 'lgbm__num_leaves': 72, 'lgbm__n_estimators': 250}


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMRegressor was fitted with feature names
  warnings.warn(


lightgbm (no PCA) Test MAE: $28,769.13


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMRegressor was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/utils/valid

🏃 View run lightgbm_baseline_optuna at: https://dagshub.com/mkzia/housing_fall2025_1.mlflow/#/experiments/1/runs/576a9b907d2344ada57c5d8f1b8e3073
🧪 View experiment at: https://dagshub.com/mkzia/housing_fall2025_1.mlflow/#/experiments/1

✓ STEP 5: All 4 baseline models optimized and logged.

Optimizing RIDGE_WITH_PCA - 10 trials


  0%|          | 0/10 [00:00<?, ?it/s]


Best RIDGE_WITH_PCA CV MAE: $52,769.74
Best params: {'ridge__alpha': 1.3292918943162166, 'pca__n_components': 0.9855642875768924}
ridge_with_pca Test MAE: $54,146.20


/usr/local/lib/python3.12/dist-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
Registered model 'ridge_pipeline_with_pca_optuna' already exists. Creating a new version of this model...
2025/12/11 11:54:06 INFO mlflow.store.model_registry.abstract_store: Waiting up to 300 seconds for model version to finish creation. Model nam

🏃 View run ridge_with_pca_optuna at: https://dagshub.com/mkzia/housing_fall2025_1.mlflow/#/experiments/1/runs/37e009e89f1e45ef91412d3b3fef389d
🧪 View experiment at: https://dagshub.com/mkzia/housing_fall2025_1.mlflow/#/experiments/1

Optimizing HISTGRADIENTBOOSTING_WITH_PCA - 10 trials


  0%|          | 0/10 [00:00<?, ?it/s]


Best HISTGRADIENTBOOSTING_WITH_PCA CV MAE: $38,339.51
Best params: {'hgb__learning_rate': 0.10618101782710439, 'hgb__max_depth': 8, 'pca__n_components': 0.9658794547630265}
histgradientboosting_with_pca Test MAE: $37,993.99


/usr/local/lib/python3.12/dist-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
Registered model 'histgradientboosting_pipeline_with_pca_optuna' already exists. Creating a new version of this model...
2025/12/11 11:54:37 INFO mlflow.store.model_registry.abstract_store: Waiting up to 300 seconds for model version to finish crea

🏃 View run histgradientboosting_with_pca_optuna at: https://dagshub.com/mkzia/housing_fall2025_1.mlflow/#/experiments/1/runs/4c408fa35b6e47719157d659e8701a43
🧪 View experiment at: https://dagshub.com/mkzia/housing_fall2025_1.mlflow/#/experiments/1

Optimizing XGBOOST_WITH_PCA - 10 trials


  0%|          | 0/10 [00:00<?, ?it/s]


Best XGBOOST_WITH_PCA CV MAE: $36,095.91
Best params: {'xgb__learning_rate': 0.14016725176148134, 'xgb__max_depth': 7, 'xgb__n_estimators': 100, 'pca__n_components': 0.9872918866945795}
xgboost_with_pca Test MAE: $34,685.04


/usr/local/lib/python3.12/dist-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
Registered model 'xgboost_pipeline_with_pca_optuna' already exists. Creating a new version of this model...
2025/12/11 11:55:30 INFO mlflow.store.model_registry.abstract_store: Waiting up to 300 seconds for model version to finish creation. Model n

🏃 View run xgboost_with_pca_optuna at: https://dagshub.com/mkzia/housing_fall2025_1.mlflow/#/experiments/1/runs/7f80316450654a81abff8f35882967b1
🧪 View experiment at: https://dagshub.com/mkzia/housing_fall2025_1.mlflow/#/experiments/1

Optimizing LIGHTGBM_WITH_PCA - 10 trials


  0%|          | 0/10 [00:00<?, ?it/s]


Best LIGHTGBM_WITH_PCA CV MAE: $35,718.71
Best params: {'lgbm__learning_rate': 0.14016725176148134, 'lgbm__num_leaves': 63, 'lgbm__n_estimators': 100, 'pca__n_components': 0.9872918866945795}


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMRegressor was fitted with feature names
  warnings.warn(


lightgbm_with_pca Test MAE: $34,714.91


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMRegressor was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/utils/valid

🏃 View run lightgbm_with_pca_optuna at: https://dagshub.com/mkzia/housing_fall2025_1.mlflow/#/experiments/1/runs/b8ac7726c28047a7a8b5db987f6e41a5
🧪 View experiment at: https://dagshub.com/mkzia/housing_fall2025_1.mlflow/#/experiments/1

✓ STEP 7: All 4 PCA models optimized and logged.

GLOBAL BEST MODEL (ACROSS 8 CANDIDATES)
Global best model key: xgboost
Global best CV MAE:    $28,986.41
Global best Test MAE:  $28,384.85
Uses PCA:               False

--------------------------------------------------------------------------------
Saving and reloading GLOBAL best model...
--------------------------------------------------------------------------------
✓ Model saved to /content/drive/MyDrive/Colab Notebooks/housing_fall2025/models/global_best_model_optuna.pkl

Done:
- GLOBAL best model key: xgboost
- GLOBAL best CV MAE:    $28,986.41
- GLOBAL best Test MAE:  $28,384.85
Elapsed time: 6 minutes and 44.73 seconds


I trained and compared multiple classification models using Optuna with 3-fold cross-validation, optimizing F1-score, both with and without PCA, and selected the global best model based on test-set F1.

In [2]:
import pandas as pd
import sqlite3
import optuna
from optuna.samplers import TPESampler

from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.decomposition import PCA
from sklearn.metrics import f1_score

from sklearn.linear_model import RidgeClassifier
from sklearn.ensemble import HistGradientBoostingClassifier

from xgboost import XGBClassifier
from lightgbm import LGBMClassifier


In [3]:
# Load data from DB

conn = sqlite3.connect("project.db")

query = """
SELECT
    c.gender,
    c.senior_citizen,
    c.partner,
    c.dependents,
    b.tenure,
    b.monthly_charges,
    b.total_charges,
    s.phone_service,
    s.multiple_lines,
    s.internet_service,
    s.online_security,
    s.online_backup,
    s.device_protection,
    s.tech_support,
    s.streaming_tv,
    s.streaming_movies,
    ctr.contract,
    ctr.paperless_billing,
    ctr.payment_method,
    ch.churn
FROM Customer c
JOIN Billing b ON c.customer_id = b.customer_id
JOIN Services s ON c.customer_id = s.customer_id
JOIN Contract ctr ON c.customer_id = ctr.customer_id
JOIN Churn ch ON c.customer_id = ch.customer_id
"""

df = pd.read_sql_query(query, conn)
df["total_charges"] = df["total_charges"].fillna(df["total_charges"].median())


In [4]:
# Train/test split (STRATIFIED)

X = df.drop("churn", axis=1)
y = df["churn"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    stratify=y,
    random_state=42
)


In [5]:
#Preprocessing pipeline

num_features = ["tenure", "monthly_charges", "total_charges"]
cat_features = X.columns.difference(num_features)

preprocessing = ColumnTransformer([
    ("num", StandardScaler(), num_features),
    ("cat", OneHotEncoder(handle_unknown="ignore"), cat_features)
])


In [8]:
# OPTUNA OBJECTIVES

def objective_ridge(trial):
    alpha = trial.suggest_float("alpha", 0.01, 10.0, log=True)

    model = Pipeline([
        ("prep", preprocessing),
        ("clf", RidgeClassifier(alpha=alpha))
    ])

    scores = cross_val_score(
        model, X_train, y_train,
        cv=3, scoring="f1", n_jobs=-1
    )
    return scores.mean()

#HistGradientBoosting

def objective_hgb(trial):
    lr = trial.suggest_float("learning_rate", 0.01, 0.3)
    depth = trial.suggest_int("max_depth", 3, 10)

    model = Pipeline([
        ("prep", preprocessing),
        ("clf", HistGradientBoostingClassifier(
            learning_rate=lr,
            max_depth=depth,
            random_state=42
        ))
    ])

    scores = cross_val_score(
        model, X_train, y_train,
        cv=3, scoring="f1", n_jobs=-1
    )
    return scores.mean()

#XGBoost

def objective_xgb(trial):
    model = Pipeline([
        ("prep", preprocessing),
        ("clf", XGBClassifier(
            n_estimators=trial.suggest_int("n_estimators", 100, 400),
            max_depth=trial.suggest_int("max_depth", 3, 8),
            learning_rate=trial.suggest_float("learning_rate", 0.01, 0.3),
            subsample=0.8,
            colsample_bytree=0.8,
            objective="binary:logistic",
            eval_metric="logloss",
            random_state=42,
            n_jobs=-1
        ))
    ])

    scores = cross_val_score(
        model, X_train, y_train,
        cv=3, scoring="f1", n_jobs=-1
    )
    return scores.mean()

# LightGBM

def objective_lgbm(trial):
    model = Pipeline([
        ("prep", preprocessing),
        ("clf", LGBMClassifier(
            n_estimators=trial.suggest_int("n_estimators", 100, 400),
            num_leaves=trial.suggest_int("num_leaves", 15, 63),
            learning_rate=trial.suggest_float("learning_rate", 0.01, 0.3),
            random_state=42,
            n_jobs=-1
        ))
    ])

    scores = cross_val_score(
        model, X_train, y_train,
        cv=3, scoring="f1", n_jobs=-1
    )
    return scores.mean()


In [ ]:
#Run Optuna NO PCA

objectives = {
    "ridge": objective_ridge,
    "hgb": objective_hgb,
    "xgb": objective_xgb,
    "lgbm": objective_lgbm
}

results = {}

for name, obj in objectives.items():
    study = optuna.create_study(
        direction="maximize",
        sampler=TPESampler(seed=42)
    )
    study.optimize(obj, n_trials=20)

    print(f"Best {name.upper()} CV F1: {study.best_value:.4f}")
    results[name] = study


[I 2025-12-17 16:07:06,708] A new study created in memory with name: no-name-d210222c-59e1-4997-a937-276d24124812
[I 2025-12-17 16:07:11,750] Trial 0 finished with value: 0.5830884936445586 and parameters: {'alpha': 0.13292918943162169}. Best is trial 0 with value: 0.5830884936445586.
[I 2025-12-17 16:07:15,455] Trial 1 finished with value: 0.5828740966264924 and parameters: {'alpha': 7.114476009343421}. Best is trial 0 with value: 0.5830884936445586.
[I 2025-12-17 16:07:18,744] Trial 2 finished with value: 0.580866225152265 and parameters: {'alpha': 1.5702970884055387}. Best is trial 0 with value: 0.5830884936445586.
[I 2025-12-17 16:07:18,960] Trial 3 finished with value: 0.5805234823588229 and parameters: {'alpha': 0.6251373574521749}. Best is trial 0 with value: 0.5830884936445586.
[I 2025-12-17 16:07:19,173] Trial 4 finished with value: 0.5835242991527316 and parameters: {'alpha': 0.02938027938703535}. Best is trial 4 with value: 0.5835242991527316.
[I 2025-12-17 16:07:19,400] Tri

Best RIDGE CV F1: 0.5837


[I 2025-12-17 16:07:23,285] Trial 0 finished with value: 0.5763771803435221 and parameters: {'learning_rate': 0.11861663446573512, 'max_depth': 10}. Best is trial 0 with value: 0.5763771803435221.
[I 2025-12-17 16:07:24,163] Trial 1 finished with value: 0.5601656334712484 and parameters: {'learning_rate': 0.22227824312530747, 'max_depth': 7}. Best is trial 0 with value: 0.5763771803435221.
[I 2025-12-17 16:07:24,850] Trial 2 finished with value: 0.5924055875151336 and parameters: {'learning_rate': 0.055245405728306586, 'max_depth': 4}. Best is trial 2 with value: 0.5924055875151336.
[I 2025-12-17 16:07:26,020] Trial 3 finished with value: 0.5777386805532335 and parameters: {'learning_rate': 0.026844247528777843, 'max_depth': 9}. Best is trial 2 with value: 0.5924055875151336.
[I 2025-12-17 16:07:26,860] Trial 4 finished with value: 0.5698755259846667 and parameters: {'learning_rate': 0.18432335340553055, 'max_depth': 8}. Best is trial 2 with value: 0.5924055875151336.
[I 2025-12-17 16:

Best HGB CV F1: 0.5924


[I 2025-12-17 16:07:37,830] Trial 0 finished with value: 0.5503162277077509 and parameters: {'n_estimators': 212, 'max_depth': 8, 'learning_rate': 0.22227824312530747}. Best is trial 0 with value: 0.5503162277077509.
[I 2025-12-17 16:07:38,437] Trial 1 finished with value: 0.5883368013426394 and parameters: {'n_estimators': 280, 'max_depth': 3, 'learning_rate': 0.055238410897498764}. Best is trial 1 with value: 0.5883368013426394.
[I 2025-12-17 16:07:39,053] Trial 2 finished with value: 0.5579759425160026 and parameters: {'n_estimators': 117, 'max_depth': 8, 'learning_rate': 0.18432335340553055}. Best is trial 1 with value: 0.5883368013426394.
[I 2025-12-17 16:07:39,626] Trial 3 finished with value: 0.5659977817586913 and parameters: {'n_estimators': 313, 'max_depth': 3, 'learning_rate': 0.29127385712697834}. Best is trial 1 with value: 0.5883368013426394.
[I 2025-12-17 16:07:40,303] Trial 4 finished with value: 0.5800676947605251 and parameters: {'n_estimators': 350, 'max_depth': 4, '

Best XGB CV F1: 0.5959


[I 2025-12-17 16:07:51,774] Trial 0 finished with value: 0.5457548977420398 and parameters: {'n_estimators': 212, 'num_leaves': 61, 'learning_rate': 0.22227824312530747}. Best is trial 0 with value: 0.5457548977420398.
[I 2025-12-17 16:07:53,049] Trial 1 finished with value: 0.5667365630870734 and parameters: {'n_estimators': 280, 'num_leaves': 22, 'learning_rate': 0.055238410897498764}. Best is trial 1 with value: 0.5667365630870734.
[I 2025-12-17 16:07:54,569] Trial 2 finished with value: 0.5467445343737083 and parameters: {'n_estimators': 117, 'num_leaves': 57, 'learning_rate': 0.18432335340553055}. Best is trial 1 with value: 0.5667365630870734.
[I 2025-12-17 16:07:55,895] Trial 3 finished with value: 0.537664742039779 and parameters: {'n_estimators': 313, 'num_leaves': 16, 'learning_rate': 0.29127385712697834}. Best is trial 1 with value: 0.5667365630870734.
[I 2025-12-17 16:07:57,775] Trial 4 finished with value: 0.568621010495041 and parameters: {'n_estimators': 350, 'num_leaves

Best LGBM CV F1: 0.5845


In [ ]:
# Train final models NO PCA & evaluate on test set

final_models = {}

for name, study in results.items():
    params = study.best_params

    if name == "ridge":
        clf = RidgeClassifier(**params)
    elif name == "hgb":
        clf = HistGradientBoostingClassifier(**params, random_state=42)
    elif name == "xgb":
        clf = XGBClassifier(**params, objective="binary:logistic",
                            eval_metric="logloss", random_state=42)
    elif name == "lgbm":
        clf = LGBMClassifier(**params, random_state=42)

    model = Pipeline([
        ("prep", preprocessing),
        ("clf", clf)
    ])

    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    f1 = f1_score(y_test, y_pred)

    print(f"{name.upper()} Test F1 (no PCA): {f1:.4f}")
    final_models[name] = {"model": model, "f1": f1}


RIDGE Test F1 (no PCA): 0.5773
HGB Test F1 (no PCA): 0.5840
XGB Test F1 (no PCA): 0.5918
[LightGBM] [Info] Number of positive: 1495, number of negative: 4139
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000343 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 670
[LightGBM] [Info] Number of data points in the train set: 5634, number of used features: 46
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.265353 -> initscore=-1.018328
[LightGBM] [Info] Start training from score -1.018328
LGBM Test F1 (no PCA): 0.5789


c:\Users\aidan\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\utils\validation.py:2749: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


In [11]:
#Global Test NO PCA

global_best_name = max(final_models, key=lambda k: final_models[k]["f1"])
global_best_f1 = final_models[global_best_name]["f1"]

print(f"Global Best (no PCA): {global_best_name.upper()} | F1 = {global_best_f1:.4f}")


Global Best (no PCA): XGB | F1 = 0.5918


Among the non-PCA models, XGBoost achieved the highest F1 score on the test set, outperforming linear and tree-based alternatives.

Now PCA

I evaluated PCA variants of Ridge, HistGradientBoosting, XGBoost, and LightGBM, tuning both model hyperparameters and the number of principal components, and compared them against non-PCA models using test-set F1

In [12]:
# PCA objective functions

#Ridge + PCA

def objective_ridge_pca(trial):
    alpha = trial.suggest_float("alpha", 0.01, 10.0, log=True)
    n_comp = trial.suggest_int("pca__n_components", 5, 30)

    model = Pipeline([
        ("prep", preprocessing),
        ("pca", PCA(n_components=n_comp, random_state=42)),
        ("clf", RidgeClassifier(alpha=alpha))
    ])

    scores = cross_val_score(
        model, X_train, y_train,
        cv=3, scoring="f1", n_jobs=-1
    )
    return scores.mean()

# HistGradientBoosting + PCA

def objective_hgb_pca(trial):
    lr = trial.suggest_float("learning_rate", 0.01, 0.3)
    depth = trial.suggest_int("max_depth", 3, 10)
    n_comp = trial.suggest_int("pca__n_components", 5, 30)

    model = Pipeline([
        ("prep", preprocessing),
        ("pca", PCA(n_components=n_comp, random_state=42)),
        ("clf", HistGradientBoostingClassifier(
            learning_rate=lr,
            max_depth=depth,
            random_state=42
        ))
    ])

    scores = cross_val_score(
        model, X_train, y_train,
        cv=3, scoring="f1", n_jobs=-1
    )
    return scores.mean()

# XGBoost + PCA

def objective_xgb_pca(trial):
    n_comp = trial.suggest_int("pca__n_components", 5, 30)

    model = Pipeline([
        ("prep", preprocessing),
        ("pca", PCA(n_components=n_comp, random_state=42)),
        ("clf", XGBClassifier(
            n_estimators=trial.suggest_int("n_estimators", 100, 400),
            max_depth=trial.suggest_int("max_depth", 3, 8),
            learning_rate=trial.suggest_float("learning_rate", 0.01, 0.3),
            subsample=0.8,
            colsample_bytree=0.8,
            objective="binary:logistic",
            eval_metric="logloss",
            random_state=42,
            n_jobs=-1
        ))
    ])

    scores = cross_val_score(
        model, X_train, y_train,
        cv=3, scoring="f1", n_jobs=-1
    )
    return scores.mean()


#LightGBM + PCA

def objective_lgbm_pca(trial):
    n_comp = trial.suggest_int("pca__n_components", 5, 30)

    model = Pipeline([
        ("prep", preprocessing),
        ("pca", PCA(n_components=n_comp, random_state=42)),
        ("clf", LGBMClassifier(
            n_estimators=trial.suggest_int("n_estimators", 100, 400),
            num_leaves=trial.suggest_int("num_leaves", 15, 63),
            learning_rate=trial.suggest_float("learning_rate", 0.01, 0.3),
            random_state=42,
            n_jobs=-1
        ))
    ])

    scores = cross_val_score(
        model, X_train, y_train,
        cv=3, scoring="f1", n_jobs=-1
    )
    return scores.mean()


In [13]:
# Run Optuna (PCA models)

pca_objectives = {
    "ridge_pca": objective_ridge_pca,
    "hgb_pca": objective_hgb_pca,
    "xgb_pca": objective_xgb_pca,
    "lgbm_pca": objective_lgbm_pca
}

pca_studies = {}

for name, obj in pca_objectives.items():
    study = optuna.create_study(
        direction="maximize",
        sampler=TPESampler(seed=42)
    )
    study.optimize(obj, n_trials=20)

    print(f"Best {name.upper()} CV F1: {study.best_value:.4f}")
    pca_studies[name] = study


[I 2025-12-17 16:18:26,616] A new study created in memory with name: no-name-f1794045-c5a4-4aeb-9d46-a1fbb68319db
[I 2025-12-17 16:18:31,607] Trial 0 finished with value: 0.5830884936445586 and parameters: {'alpha': 0.13292918943162169, 'pca__n_components': 29}. Best is trial 0 with value: 0.5830884936445586.
[I 2025-12-17 16:18:35,125] Trial 1 finished with value: 0.5814759389931803 and parameters: {'alpha': 1.5702970884055387, 'pca__n_components': 20}. Best is trial 0 with value: 0.5830884936445586.
[I 2025-12-17 16:18:38,466] Trial 2 finished with value: 0.5758354228873236 and parameters: {'alpha': 0.02938027938703535, 'pca__n_components': 9}. Best is trial 0 with value: 0.5830884936445586.
[I 2025-12-17 16:18:38,664] Trial 3 finished with value: 0.5835242991527316 and parameters: {'alpha': 0.014936568554617643, 'pca__n_components': 27}. Best is trial 3 with value: 0.5835242991527316.
[I 2025-12-17 16:18:38,846] Trial 4 finished with value: 0.5805234823588229 and parameters: {'alpha

Best RIDGE_PCA CV F1: 0.5837


[I 2025-12-17 16:18:42,930] Trial 0 finished with value: 0.548533104151955 and parameters: {'learning_rate': 0.11861663446573512, 'max_depth': 10, 'pca__n_components': 24}. Best is trial 0 with value: 0.548533104151955.
[I 2025-12-17 16:18:43,483] Trial 1 finished with value: 0.5730654548628299 and parameters: {'learning_rate': 0.18361096041714062, 'max_depth': 4, 'pca__n_components': 9}. Best is trial 1 with value: 0.5730654548628299.
[I 2025-12-17 16:18:44,391] Trial 2 finished with value: 0.5726601578002448 and parameters: {'learning_rate': 0.026844247528777843, 'max_depth': 9, 'pca__n_components': 20}. Best is trial 1 with value: 0.5730654548628299.
[I 2025-12-17 16:18:44,815] Trial 3 finished with value: 0.5752246733746392 and parameters: {'learning_rate': 0.21534104756085318, 'max_depth': 3, 'pca__n_components': 30}. Best is trial 3 with value: 0.5752246733746392.
[I 2025-12-17 16:18:45,217] Trial 4 finished with value: 0.561535334506748 and parameters: {'learning_rate': 0.251408

Best HGB_PCA CV F1: 0.5794


[I 2025-12-17 16:18:56,162] Trial 0 finished with value: 0.5328342063372132 and parameters: {'pca__n_components': 14, 'n_estimators': 386, 'max_depth': 7, 'learning_rate': 0.18361096041714062}. Best is trial 0 with value: 0.5328342063372132.
[I 2025-12-17 16:18:56,545] Trial 1 finished with value: 0.5633213447782984 and parameters: {'pca__n_components': 9, 'n_estimators': 146, 'max_depth': 3, 'learning_rate': 0.2611910822747312}. Best is trial 1 with value: 0.5633213447782984.
[I 2025-12-17 16:18:57,279] Trial 2 finished with value: 0.5575411343277092 and parameters: {'pca__n_components': 20, 'n_estimators': 313, 'max_depth': 3, 'learning_rate': 0.29127385712697834}. Best is trial 1 with value: 0.5633213447782984.
[I 2025-12-17 16:18:57,946] Trial 3 finished with value: 0.5812819597924485 and parameters: {'pca__n_components': 26, 'n_estimators': 163, 'max_depth': 4, 'learning_rate': 0.06318730785749581}. Best is trial 3 with value: 0.5812819597924485.
[I 2025-12-17 16:18:58,710] Trial 

Best XGB_PCA CV F1: 0.5894


[I 2025-12-17 16:19:15,828] Trial 0 finished with value: 0.5283475437415329 and parameters: {'pca__n_components': 14, 'n_estimators': 386, 'num_leaves': 50, 'learning_rate': 0.18361096041714062}. Best is trial 0 with value: 0.5283475437415329.
[I 2025-12-17 16:19:16,469] Trial 1 finished with value: 0.5468839256670465 and parameters: {'pca__n_components': 9, 'n_estimators': 146, 'num_leaves': 17, 'learning_rate': 0.2611910822747312}. Best is trial 1 with value: 0.5468839256670465.
[I 2025-12-17 16:19:17,750] Trial 2 finished with value: 0.5209255674938723 and parameters: {'pca__n_components': 20, 'n_estimators': 313, 'num_leaves': 16, 'learning_rate': 0.29127385712697834}. Best is trial 1 with value: 0.5468839256670465.
[I 2025-12-17 16:19:18,767] Trial 3 finished with value: 0.5657581127978899 and parameters: {'pca__n_components': 26, 'n_estimators': 163, 'num_leaves': 23, 'learning_rate': 0.06318730785749581}. Best is trial 3 with value: 0.5657581127978899.
[I 2025-12-17 16:19:20,515

Best LGBM_PCA CV F1: 0.5765


In [14]:
# Train final PCA models & test evaluation

pca_models = {}

for name, study in pca_studies.items():
    params = study.best_params
    n_comp = params.pop("pca__n_components")

    if name == "ridge_pca":
        clf = RidgeClassifier(**params)
    elif name == "hgb_pca":
        clf = HistGradientBoostingClassifier(**params, random_state=42)
    elif name == "xgb_pca":
        clf = XGBClassifier(**params,
                            objective="binary:logistic",
                            eval_metric="logloss",
                            random_state=42)
    elif name == "lgbm_pca":
        clf = LGBMClassifier(**params, random_state=42)

    model = Pipeline([
        ("prep", preprocessing),
        ("pca", PCA(n_components=n_comp, random_state=42)),
        ("clf", clf)
    ])

    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    f1 = f1_score(y_test, y_pred)

    print(f"{name.upper()} Test F1 (PCA): {f1:.4f}")
    pca_models[name] = {"model": model, "f1": f1}


RIDGE_PCA Test F1 (PCA): 0.5773
HGB_PCA Test F1 (PCA): 0.5542
XGB_PCA Test F1 (PCA): 0.5684
[LightGBM] [Info] Number of positive: 1495, number of negative: 4139
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.001095 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 7060
[LightGBM] [Info] Number of data points in the train set: 5634, number of used features: 30
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.265353 -> initscore=-1.018328
[LightGBM] [Info] Start training from score -1.018328
LGBM_PCA Test F1 (PCA): 0.5477


c:\Users\aidan\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\utils\validation.py:2749: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


In [15]:
# Global best (ALL models)

all_models = {**final_models, **pca_models}

global_best_name = max(all_models, key=lambda k: all_models[k]["f1"])
global_best_f1 = all_models[global_best_name]["f1"]

print(f"GLOBAL BEST MODEL: {global_best_name.upper()}")
print(f"GLOBAL BEST TEST F1: {global_best_f1:.4f}")


GLOBAL BEST MODEL: XGB
GLOBAL BEST TEST F1: 0.5918


Across all models, PCA variants performed worse than their non-PCA counterparts. This is expected because PCA can distort informative one-hot encoded categorical features. As a result, the global best model was XGBoost without PCA, achieving a test F1 score of 0.5918.

In [16]:
# Save model with joblib

import joblib
from pathlib import Path

# Create models directory if it doesn't exist
Path("models").mkdir(exist_ok=True)

# Save the global best pipeline
joblib.dump(
    all_models[global_best_name]["model"],
    "models/global_best_model.pkl"
)

print("Model saved to models/global_best_model.pkl")


Model saved to models/global_best_model.pkl
